crates/bws/src/command/run.rs

```diff
@@ -0,0 +1,135 @@
+use std::{io::Read, process};
+
+use atty::Stream;
+use bitwarden::{
+    secrets_manager::{
+        secrets::{SecretIdentifiersByProjectRequest, SecretIdentifiersRequest, SecretsGetRequest},
+        ClientSecretsExt,
+    },
+    Client,
+};
+use clap::CommandFactory;
+use color_eyre::eyre::Result;
+use uuid::Uuid;
+use which::which;
+
+use crate::{
+    util::{is_valid_posix_name, uuid_to_posix},
+    Cli, ACCESS_TOKEN_KEY_VAR_NAME,
+};
+
+pub(crate) async fn run(
+    client: Client,
+    organization_id: Uuid,
+    project_id: Option<Uuid>,
+    uuids_as_keynames: bool,
+    no_inherit_env: bool,
+    shell: Option<String>,
+    command: Vec<String>,
+) -> Result<()> {
+    let shell = match std::env::consts::OS {
+        "windows" => shell.unwrap_or_else(|| "powershell".to_string()),
+        _ => shell.unwrap_or_else(|| "sh".to_string()),
+    };
+
+    if which(&shell).is_err() {
+        eprintln!("Error: shell '{}' not found", shell);
+        std::process::exit(1);
+    }
+
+    let user_command = if command.is_empty() {
+        if atty::is(Stream::Stdin) {
+            eprintln!("{}", Cli::command().render_help().ansi());
+            std::process::exit(1);
+        }
+
+        let mut buffer = String::new();
+        std::io::stdin().read_to_string(&mut buffer)?;
+        buffer
+    } else {
+        command.join(" ")
+    };
+
+    if user_command.is_empty() {
+        let mut cmd = Cli::command();
+        eprintln!("{}", cmd.render_help().ansi());
+        std::process::exit(1);
+    }
+
+    let res = if let Some(project_id) = project_id {
+        client
+            .secrets()
+            .list_by_project(&SecretIdentifiersByProjectRequest { project_id })
+            .await?
+    } else {
+        client
+            .secrets()
+            .list(&SecretIdentifiersRequest { organization_id })
+            .await?
+    };
+
+    let secret_ids = res.data.into_iter().map(|e| e.id).collect();
+    let secrets = client
+        .secrets()
+        .get_by_ids(SecretsGetRequest { ids: secret_ids })
+        .await?
+        .data;
+
+    let mut environment = std::collections::HashMap::new();
+
+    secrets.into_iter().for_each(|s| {
+        let key = if uuids_as_keynames {
+            uuid_to_posix(&s.id)
+        } else {
+            s.key
+        };
+
+        if !is_valid_posix_name(&key) {
+            eprintln!(
+                "Warning: secret '{}' does not have a POSIX-compliant name",
+                key
+            );
+        }
+
+        match environment.contains_key(&key) {
+            true => {
+                eprintln!(
+                    "Error: multiple secrets with name '{}' found. Use --uuids-as-keynames or use unique names for secrets.",
+                    key
+                );
+                std::process::exit(1);
+            }
+            false => {
+                environment.insert(key, s.value);
+            }
+        }
+    });
```

Comment:
**suggestion:** It's generally more idiomatic to use map or flatmap and collect.

This could look something like

```rust
let evironment = secres
    .into_iter()
    .map(|s| {
        if uuids_as_keynames {
            uuid_to_posix(&s.id)
        } else {
            s.key
        }
    })
    .inspect(|s| {
        if !is_valid_posix_name(&key) {
            eprintln!(
                "Warning: secret '{}' does not have a POSIX-compliant name",
                key
            );
        }
    });

// Using itertools https://docs.rs/itertools/latest/itertools/trait.Itertools.html#method.all_unique
if !environment.all_unique() {
    return Err("Not unique)
}
```

Reply:
An attempt at making this portion more idiomatic was made in 3fcd32ce1d7ba50da19d4e751dd2921a013912c1, with a follow-up in eda00aa19bf8ec06498932224c976f855abae526 (in response to a clippy warning).

I moved things around a little bit so we could exit earlier if we detect duplicates. Let me know if that makes sense.